### PR TITLE
Retry recycle or delete operation on failure.

### PR DESCRIPTION
**Pull request by jsafrane**
Recycle controller tries to recycle or delete a PV several times.
It stores count of failed attempts and timestamp of the last attempt in
~~annotations of the PV~~ internal map.

By default, the controller tries to recycle/delete a PV 3 times in
10 minutes interval. These values are configurable by
kube-controller-manager --pvrecycler-retry-count=X --pvclaimbinder-sync-period=Y
arguments.

cc: @kubernetes/rh-storage

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
@markturansky - can you pick this up?

**Comment by markturansky** [comment]
race condition.  i assigned myself just as you posted.

yes, i will review this.

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 1]
```diff
@@ -81,6 +81,7 @@ type CMServer struct {
 	ResourceQuotaSyncPeriod           time.Duration
 	NamespaceSyncPeriod               time.Duration
 	PVClaimBinderSyncPeriod           time.Duration
+	PVRecycleRetryCount               int
```
Comment (markturansky) [inline cmd/kube-controller-manager/app/controllermanager.go]: I think this is better in the VolumeConfigFlags struct.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 2]
```diff
@@ -94,16 +116,61 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
...
+	// The volume is Failed, should we retry recycling?
+	retryCount := recycler.getRetryCount(pv)
+	glog.V(5).Infof("PersistentVolume[%s] %d attempts failed so far", pv.Name, retryCount)
+	if retryCount >= recycler.maximumRetryCount {
+		glog.V(5).Infof("PersistentVolume[%s] skipping deletion, %d attempts failed", pv.Name, retryCount)
+		return fmt.Errorf("%d attempts failed", retryCount)
+	}
+
+	// retry-count is fine, check the timestamp
+	lastFailureStr, ok := pv.Annotations[pvRecycleLastFailureAnnotationKey]
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: what are the benefits/consequences of using annotations to store this versus a local map?

A local map containing counts would remove annotation keys, remove PV updates, etc. but it wouldn't guarantee 3 retries.  A restarted controller manager would necessarily have an empty map and the PV could go 3 more times, even if it attempted recycling twice before the controller manager was restarted.

I don't think the guarantee of X times is important, especially if it makes the code smaller/easier for this feature.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 1 above)
Comment (jsafrane) [inline cmd/kube-controller-manager/app/controllermanager.go]: Ok, moved.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 2 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Local cache is probably easier to handle and does not generate 'pv updated' events to other controllers, which is much better. I'm adding a patch with local cache of failed volumes (to be squashed into the final patch eventually).

On the other side, with the local cache the state is hidden inside the controller and any debugging/troubleshooting is much harder (as I found out right now). 
Annotations in `kubectl get pv -o yaml` were quite useful.

**Comment by jsafrane** [comment]
@k8s-bot unit test this

File: cmd/kube-controller-manager/app/options/options.go [hunk 3]
```diff
@@ -85,6 +85,7 @@ type VolumeConfigFlags struct {
 	PersistentVolumeRecyclerMinimumTimeoutHostPath      int
 	PersistentVolumeRecyclerIncrementTimeoutHostPath    int
 	EnableHostPathProvisioning                          bool
+	PVRecycleRetryCount                                 int
```
Comment (markturansky) [inline cmd/kube-controller-manager/app/options/options.go]: For consistency, I think this should be `PersistentVolumeRecyclerRetryCount`, which would match the other variables in this struct.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 4]
```diff
@@ -36,25 +36,60 @@ import (
...
+	// operation fails with timestamp of the last recycle or delete failure.
+	// We must keep the timestamp to make sure we retry the operation only after
+	// some time (PersistentVolumeRecycler.syncPeriod). Without checking the
+	// timestamp the controller would enter endless loop retrying the operation
+	// -  a failed retry updates volume's pvRecycleRetryCountAnnotationKey with
+	// new retry count, which generates new 'volume updated' event, which
+	// triggers new retry, which is likely to fail again.
+	pvRecycleLastFailureAnnotationKey = "volume.experimental.kubernetes.io/last-failure"
+)
+
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: These are no longer used/needed if you've got a local map storing state.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 5]
```diff
@@ -94,16 +136,56 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
...
 		glog.V(5).Infof("Reclaiming PersistentVolume[%s]\n", pv.Name)
 
 		latest, err := recycler.client.GetPersistentVolume(pv.Name)
 		if err != nil {
 			return fmt.Errorf("Could not find PersistentVolume %s", pv.Name)
 		}
-		if latest.Status.Phase != api.VolumeReleased {
-			return fmt.Errorf("PersistentVolume[%s] phase is %s, expected %s.  Skipping.", pv.Name, latest.Status.Phase, api.VolumeReleased)
+
+		if err = recycler.shouldRecycle(latest); err != nil {
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Why is this here as well as on line 179 above?  This check is gated by the 1st `shouldRecycle` just a few lines earlier.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 6]
```diff
@@ -123,10 +205,32 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
 			glog.Errorf(errMsg)
 			return fmt.Errorf(errMsg)
 		}
+	} else {
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: I think you can drop `else

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 7]
```diff
@@ -151,10 +255,14 @@ func (recycler *PersistentVolumeRecycler) handleRecycle(pv *api.PersistentVolume
 		// blocks until completion
 		if err := volRecycler.Recycle(); err != nil {
 			glog.Errorf("PersistentVolume[%s] failed recycling: %+v", pv.Name, err)
+			recycler.updateFailedVolume(pv)
 			pv.Status.Message = fmt.Sprintf("Recycling error: %s", err)
 			nextPhase = api.VolumeFailed
 		} else {
 			glog.V(5).Infof("PersistentVolume[%s] successfully recycled\n", pv.Name)
+			// The volume has been recycled. Remove any retry annotations to make
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: comment needs updating.  no more annotations for state.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 8]
```diff
@@ -123,10 +205,32 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
...
+		// First failure, set retryCount to 0 (will be inceremented few lines below)
+		status = failedVolumeStatus{}
+	}
+	status.retryCount += 1
+	status.lastAttempt = time.Now()
+	recycler.failedVolumes[pv.Name] = status
+}
+
+func (recycler *PersistentVolumeRecycler) removeFailedVolume(pv *api.PersistentVolume) {
+	delete(recycler.failedVolumes, pv.Name)
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Do we need `removeFailedVolume`?  This is a one-liner and so is the `deleteVolume` func.  Maybe just `deleteVolume`?

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 5 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Well... why it was there in the original code? I did not change its logic.

Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: `shouldRecycle` is a new helper function you added to support retries.  Current `reclaimVolume` func does not contain it.

File: cmd/kube-controller-manager/app/options/options.go (hunk 3 above)
Comment (jsafrane) [inline cmd/kube-controller-manager/app/options/options.go]: Ok

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 4 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: ok

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 5 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: If you look at `reclaimVolume()` before my patch, you see that the `pv.Status.Phase == api.VolumeReleased` is checked twice too. Now that the check is moved to `shouldRecycle` and extended.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 6 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: logging in the `else` branch has proven to be valuable during testing, I'd like to leave as it is.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 8 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Good point, `deleteVolume` removed

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 7 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Good catch!

**Comment by jsafrane** [comment]
@k8s-bot test this please

**Comment by jsafrane** [comment]
@k8s-bot test this please

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 5 above)
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: You're right.   This line makes no sense, as it will never be true:

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/persistentvolume/persistentvolume_recycler_controller.go#L105

Unless I'm missing something, we don't need a 2nd check.  The 1st check can be broken in 2:

```
if phase != released, return error
if claimRef == nil, return error
...
the rest of the code...
```

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 9]
```diff
@@ -123,10 +188,28 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
 			glog.Errorf(errMsg)
 			return fmt.Errorf(errMsg)
 		}
+		return nil
+	} else {
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: By drop `else`, I mean you can remove the else clause but keep the glog statement.  This can be flattened by un-indenting and then return nil.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 5 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: There is `pv = GetPersistentVolume(pv.Name)`  before the second checks, so I just guess you wanted double-check the PV did not change. Some other controller might modify the PV (it's quite busy object these days :-).

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 9 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Oh, ok, that makes sense.

**Comment by jsafrane** [comment]
rebased

**Comment by jsafrane** [comment]
@k8s-bot unit test this

**Comment by markturansky** [comment]
All green, Jan.  I'll do another detailed review today, now that the merge queue is unblocked.

**Comment by jsafrane** [comment]
I forgot to push the 2nd patch when rebasing, pushed now.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 10]
```diff
@@ -89,6 +107,13 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
 				}
 				recycler.reclaimVolume(pv)
 			},
+			DeleteFunc: func(obj interface{}) {
+				pv, ok := obj.(*api.PersistentVolume)
+				if !ok {
+					glog.Errorf("Error casting object to PersistentVolume: %v", obj)
+				}
+				recycler.removeFailedVolume(pv)
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Would need to return if !ok.   This function would fail if pv == DeletedFinalStateUnknown.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 11]
```diff
@@ -96,16 +121,57 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
 	return recycler, nil
 }
 
+func (recycler *PersistentVolumeRecycler) shouldRecycle(pv *api.PersistentVolume) error {
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: please add a comment for what the func does.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 12]
```diff
@@ -96,16 +121,57 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
...
+	if pv.Spec.ClaimRef == nil {
+		return fmt.Errorf("Volume does not have a reference to claim")
+	}
+	if pv.Status.Phase == api.VolumeReleased {
+		// This is the first attempt to recycle, allow.
+		return nil
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: I think this is confusing now.  I understand why you've done this, because the recycler will mark the PV failed if recycling fails and you check for that status below.   But that now means 2 different states represent a need to recycle:  status=Released || status=Failed && !exceedFailedCount.  

`kubectl get pv` would show PVs that failed, but only some are truly Failed.  

We should use Released with the count you're keeping.  When the count is exceeded, the PV is truly failed.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 13]
```diff
@@ -40,21 +40,39 @@ var _ volume.VolumeHost = &PersistentVolumeRecycler{}
...
+// The controller re-tries the opration several times and it stores retry
+// count + timestamp of the last attempt here.
+type failedVolumeStatus struct {
+	// How many recycle/delete operations failed.
+	retryCount int
+	// Timestamp of the last attempt.
+	lastAttempt time.Time
 }
 
 // PersistentVolumeRecycler creates a new PersistentVolumeRecycler
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: this is my typo in the original, but since you've got this PR open, would you fix the comment to start with the correct func name?

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 5 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: I removed this double check in the last patch, I'll reload `pv` from the server and check it only once.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 11 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Added

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 12 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: I rewrote the patch to keep the volume `Released` and mark it as `Failed` only when all attempts to recycle/delete volume fail.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 10 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Fixed

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 13 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Fixed

**Comment by jsafrane** [comment]
Updated with @markturansky's comments and rebased.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 14]
```diff
@@ -40,21 +40,39 @@ var _ volume.VolumeHost = &PersistentVolumeRecycler{}
...
+	syncPeriod        time.Duration
+	// Local cache of failed recycle / delete operations. Map volume.Name -> status of the volume.
+	// Only PVs in Released state have an entry here.
+	releasedVolumes map[string]releasedVolumeStatus
 }
 
-// PersistentVolumeRecycler creates a new PersistentVolumeRecycler
-func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Duration, plugins []volume.VolumePlugin, cloud cloudprovider.Interface) (*PersistentVolumeRecycler, error) {
+// releasedVolumeStatus holds state of failed delete/recycle operation on a
+// volume. The controller re-tries the opration several times and it stores
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: typo: opration

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 15]
```diff
@@ -125,10 +184,41 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
...
+	if !found {
+		// First failure, set retryCount to 0 (will be inceremented few lines below)
+		status = releasedVolumeStatus{}
+	}
+	status.retryCount += 1
+
+	if status.retryCount > recycler.maximumRetryCount {
+		// This was the last attempt. Remove any internal state and mark the
+		// volume as Failed.
+		glog.V(3).Infof("PersistentVolume[%s] failed %dth time - marking Failed", pv.Name, status.retryCount)
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: "failed %d times" otherwise the log would read 2th or 3th.

**Comment by markturansky** [comment]
LGTM.

Over to @swagiaal for a 2nd opinion.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 16]
```diff
@@ -40,21 +40,39 @@ var _ volume.VolumeHost = &PersistentVolumeRecycler{}
...
+	stopChannel       chan struct{}
+	client            recyclerClient
+	kubeClient        client.Interface
+	pluginMgr         volume.VolumePluginMgr
+	cloud             cloudprovider.Interface
+	maximumRetryCount int
+	syncPeriod        time.Duration
+	// Local cache of failed recycle / delete operations. Map volume.Name -> status of the volume.
+	// Only PVs in Released state have an entry here.
+	releasedVolumes map[string]releasedVolumeStatus
```
Comment (swagiaal) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Hmm.. I am not too comfortable with this approach. Wouldn't the volume state be better stored as an annotation on the PV ? You should be able to look in one place and figure out the sate of a PV. This makes it easier people other than the original author to debug volume problems. In addition if we end up using the State design pattern to clean volume states up we'll have to undo this and put it back in the volume.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 17]
```diff
@@ -125,10 +184,41 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
 			glog.Errorf(errMsg)
 			return fmt.Errorf(errMsg)
 		}
+		return nil
 	}
+	glog.V(3).Infof("PersistentVolume[%s] phase %s - skipping: %v", pv.Name, pv.Status.Phase, err)
 	return nil
 }
 
+// updateReleasedVolume evaluates a failed Recycle/Delete operation, updates
```
Comment (swagiaal) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: s/updateReleasedVolume/handleReleaseFailure

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 16 above)
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: See buried comment above https://github.com/kubernetes/kubernetes/pull/19365#discussion_r49107689

Using PV annotations generated many events as the PV was saved over and over. Also require more keys.  Code was smaller after switching to local state.  In the case of controller restart, all PVs start at 0, so the MaxRetries var is not guaranteed (it would be w/ PV annotation), but that doesn't seem important.

**Comment by markturansky** [comment]
Requires rebase

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 18]
```diff
@@ -125,10 +184,41 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
...
+// internal controller state with new nr. of attempts and timestamp of the last
+// attempt. Based of nr. of failures it returns the next state of the volume
```
Comment (swagiaal) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: s/Based of nr./Based on the number/

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 15 above)
Comment (swagiaal) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: This has not been addressed it looks like

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 17 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Fixed

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 18 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Fixed

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 14 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: fixed

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 15 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: Fixed

**Comment by jsafrane** [comment]
Rebased and fixed reported issues.

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go (hunk 16 above)
Comment (swagiaal) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: I see... as @jsafrane has noted already though we'll pay the price of having scattered state. I wont block this though since you and Jan are aware of the trade off.

**Comment by swagiaal** [comment]
One nit otherwise LGTM
@markturansky @jsafrane

**Comment by jsafrane** [comment]
As @swagiaal 
* all variables `.*RetryCount` to `.*MaximumRetry`
* cmdline option `--pv-recycler-retry-count` to `pv-recycler-maximum-retry`

**Comment by jsafrane** [comment]
Rebased (loosing LGTM again).

**Comment by markturansky** [comment]
Re-adding LGTM after rebase.

**Comment by jsafrane** [comment]
@markturansky, I had to regenerate docs, please review again

File: pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 19]
```diff
@@ -46,15 +46,33 @@ type PersistentVolumeRecycler struct {
...
-// PersistentVolumeRecycler creates a new PersistentVolumeRecycler
-func NewPersistentVolumeRecycler(kubeClient clientset.Interface, syncPeriod time.Duration, plugins []volume.VolumePlugin, cloud cloudprovider.Interface) (*PersistentVolumeRecycler, error) {
+// releasedVolumeStatus holds state of failed delete/recycle operation on a
+// volume. The controller re-tries the operation several times and it stores
+// retry count + timestamp of the last attempt here.
+type releasedVolumeStatus struct {
+	// How many recycle/delete operations failed.
+	retryCount int
+	// Timestamp of the last attempt.
+	lastAttempt time.Time
```
Comment (markturansky) [inline pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: It's interesting now that we're thinking about Conditions.  These are the kinds of fileds on a Condition (type, timestamp, etc).

**Comment by markturansky** [comment]
re-added LGTM after doc regen.
